### PR TITLE
Lock pages when they are scheduled for publishing

### PR DESCRIPTION
**Pull request by kaedroho**
This is to help make the UX of scheduled pages a bit nicer. The version of the page that's in draft is always the version that will be published by scheduled publishing.

### CODE REVIEW

File: wagtail/admin/views/pages/edit.py [hunk 1]
```diff
@@ -375,79 +379,38 @@ def dispatch(self, request, page_id):
...
-                    workflow_info = format_html(
-                        _(
-                            "This page is awaiting <b>'{}'</b> in the <b>'{}'</b> workflow."
+                if (
+                    isinstance(self.lock, ScheduledForPublishLock)
+                    and self.page_perms.can_unschedule()
+                ):
+                    scheduled_revision = self.page.revisions.defer().get(
+                        approved_go_live_at__isnull=False
+                    )
```
Comment (laymonage) [inline wagtail/admin/views/pages/edit.py:399]: FYI I was able to reduce a considerable amount of duplicated queries (around 5) by turning this into a cached property in `DraftStateMixin`, and using that here and in `get_message()`, which I think should be safe in these cases.

See https://github.com/wagtail/wagtail/pull/8901/commits/01711e631df64a23a82e5364786420bb2efa3914

File: wagtail/admin/tests/pages/test_edit_page.py [hunk 2]
```diff
@@ -1278,7 +1278,7 @@ def test_edit_after_change_language_code_without_revisions(self):
         )
 
     def test_page_edit_num_queries(self):
-        with self.assertNumQueries(40):
+        with self.assertNumQueries(46):
```
Comment (laymonage) [inline wagtail/admin/tests/pages/test_edit_page.py:1281]: With the above optimisation, we should be able to bring this to 41 (or maybe even keep it at 40 in this case).

File: wagtail/admin/action_menu.py [hunk 3]
```diff
@@ -44,11 +44,10 @@ def is_shown(self, context):
             'user_page_permissions' = a UserPagePermissionsProxy for the current user, to test permissions against
             may also contain:
             'user_page_permissions_tester' = a PagePermissionTester for the current user and page
+            'lock' = a Lock object if the page is locked, otherwise None
+            'locked_for_user' = True if the lock prevents the current user from editing the page
```
Comment (gasman) [inline wagtail/admin/action_menu.py:48]: Do we guarantee that the context will always contain these variables? If so, they should be listed before "may also contain". If not, the code below needs to guard against them being missing with `context.get(...)`.

File: wagtail/locks.py [hunk 4]
```diff
@@ -102,3 +103,32 @@ def get_message(self, user):
...
+    def for_user(self, user):
+        return True
+
+    def get_message(self, user):
+        scheduled_revision = self.page.revisions.filter(
+            approved_go_live_at__isnull=False
+        ).first()
+
+        return format_html(
+            _("Page '{}' is locked and has been scheduled to go-live at {}"),
```
Comment (gasman) [inline wagtail/locks.py:128]: "go-live" shouldn't have a hyphen when used as a verb - it should be "go live".

**Review by gasman — CHANGES_REQUESTED** [review]
Very happy with the approach here! The introduction of Lock objects is a really neat way to connect business logic to the front-end messaging without things getting spaghetti-ish, and the reduction in queries is a really nice bonus too. Just some very minor niggles to sort.

**Comment by gasman** [comment]
Merged (with suggested fixes applied) in e5dd080fce3cda19e61ce5e27adc436d334d747b.

**Pull request closed without merging**